In [22]:
import nltk
nltk.download('all')
from nltk.book import *

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /Users/aya/nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to /Users/aya/nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /Users/aya/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /Users/aya/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_eng is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /Users/aya/nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package ave

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


When you run `text1.similar("monstrous")`, NLTK looks through the text (in this case, likely the text text1, which is usually "Moby Dick") to see where the word "monstrous" appears. It then checks the words surrounding "monstrous" in these contexts and finds other words that tend to appear in similar contexts (before and after the word). This process doesn’t necessarily mean that these words have the same meaning but rather that they are used in similar sentence structures or situations.

In [35]:
print("Text 1, Concordance: ")
text1.concordance("monstrous")
print("Text 1, Similar words: ")
text1.similar("monstrous")
print("Text 1, Common contexts: ")
text1.common_contexts(["delightfully", "monstrous"])

Text 1, Concordance: 
Displaying 11 of 11 matches:
ong the former , one was of a most monstrous size . ... This came towards us , 
ON OF THE PSALMS . " Touching that monstrous bulk of the whale or ork we have r
ll over with a heathenish array of monstrous clubs and spears . Some were thick
d as you gazed , and wondered what monstrous cannibal and savage could ever hav
that has survived the flood ; most monstrous and most mountainous ! That Himmal
they might scout at Moby Dick as a monstrous fable , or still worse and more de
th of Radney .'" CHAPTER 55 Of the Monstrous Pictures of Whales . I shall ere l
ing Scenes . In connexion with the monstrous pictures of whales , I am strongly
ere to enter upon those still more monstrous stories of them which are to be fo
ght have been rummaged out of this monstrous cabinet there is no telling . But 
of Whale - Bones ; for Whales of a monstrous size are oftentimes cast up dead u
Text 1, Similar words: 
true contemptible christian abundant few part

In [38]:
text3.generate()

laid by her , and said unto Cain , Where art thou , and said , Go to ,
I will not do it for ten ' s sons ; we dreamed each man according to
their generatio the firstborn said unto Laban , Because I said , Nay ,
but Sarah shall her name be . , duke Elah , duke Shobal , and Akan .
and looked upon my affliction . Bashemath Ishmael ' s blood , but Isra
for as a prince hast thou found of all the cattle in the valley , and
the wo The


"laid by her , and said unto Cain , Where art thou , and said , Go to ,\nI will not do it for ten ' s sons ; we dreamed each man according to\ntheir generatio the firstborn said unto Laban , Because I said , Nay ,\nbut Sarah shall her name be . , duke Elah , duke Shobal , and Akan .\nand looked upon my affliction . Bashemath Ishmael ' s blood , but Isra\nfor as a prince hast thou found of all the cattle in the valley , and\nthe wo The"

In [45]:
len(text3)

44764

A token is any sequence of characters that forms a meaningful unit in the text, like a word, a punctuation mark, or even an emoji (e.g., to, be, hairy, or :)).
In the example, the text "Genesis" has 44,764 tokens, meaning the entire text contains that many individual pieces (words, punctuation, etc.). Example: The phrase "to be or not to be" has 6 tokens (because we count every word separately and repeated words like "to" are counted each time they appear).

In [47]:
sorted(set(text3))
len(set(text3))

2789

A type is a unique token. It refers to the distinct vocabulary items in a text, where duplicates are ignored.
For example, in the phrase "to be or not to be," the types would be 4 distinct words: to, be, or, and not. Even though "to" and "be" appear twice, they only count once each in the vocabulary. The book of Genesis has 2,789 distinct words (types), meaning out of the 44,764 total tokens, there are only 2,789 unique words or punctuation symbols. set(text3) takes the text (which is a sequence of tokens) and converts it into a set, a collection where each item is unique. This means all duplicate tokens are removed. For example, if "to" appears multiple times in the text, it only appears once in the set. This set contains all the unique words and punctuation marks in the text.

## Lexical richness
Lexical richness is a measure of how diverse the vocabulary of a text is. In this case, the distinct words (2,789) are only 6% of the total tokens (44,764). This means that most words in the text are repeated many times. You can also say that each unique word is used 16 times on average (because 44,764 total tokens / 2,789 unique tokens ≈ 16).

In [50]:
print(len(set(text3)) / len(text3))
print(len(text3) / len(set(text3)))

0.06230453042623537
16.050197203298673


Next, let's focus on particular words. We can count how often a word occurs in a text, and compute what percentage of the text is taken up by a specific word:

In [51]:
print(text3.count("smote"))
print(100 * text4.count('a') / len(text4))

5
1.457806031353621
